### PR TITLE
Create modal launcher

### PR DESCRIPTION
**Pull request by debsmita1**
## Fixes 
https://issues.redhat.com/browse/HAC-1238

## Description
- created modal launcher using PF's modal
- removed the build log viewer modal hack to use the modal launcher

## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 
ml

## Browser conformance: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: src/shared/components/modal/ModalProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  setProps?: (children: React.ReactElement) => void;
+};
+
+export const ModalContext = React.createContext<ModalContextType>({});
+export const ModalProvider: React.FC<ModalContextType> = ({ children }) => {
+  const [context, setContext] = React.useState<React.ReactElement>();
+
+  React.useEffect(() => {
+    setContext(null);
+  }, []);
```
Comment (sahil143) [inline src/shared/components/modal/ModalProvider.tsx:15]: Why is this needed?

File: src/shared/components/modal/ModalProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { ModalPortal } from '../../../ModalPortal';
+
+type ModalContextType = {
+  props?: React.ReactElement;
+  setProps?: (children: React.ReactElement) => void;
+};
```
Comment (sahil143) [inline src/shared/components/modal/ModalProvider.tsx:7]: Why do you need to pass both context and setcontext to ModalContext

File: src/shared/components/modal/ModalProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  const [context, setContext] = React.useState<React.ReactElement>();
+
+  React.useEffect(() => {
+    setContext(null);
+  }, []);
+
+  return (
+    <>
+      <ModalPortal />
+      <ModalContext.Provider value={{ props: context, setProps: setContext }}>
```
Comment (sahil143) [inline src/shared/components/modal/ModalProvider.tsx:20]: this is misleading, you are naming context => props. And you don't need to pass context to the provider. the provider just needs to have a callback that can open the modal

File: src/shared/components/modal/ModalProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import { ModalPortal } from '../../../ModalPortal';
+
+type ModalContextType = {
+  props?: React.ReactElement;
+  setProps?: (children: React.ReactElement) => void;
+};
+
+export const ModalContext = React.createContext<ModalContextType>({});
+export const ModalProvider: React.FC<ModalContextType> = ({ children }) => {
+  const [context, setContext] = React.useState<React.ReactElement>();
```
Comment (sahil143) [inline src/shared/components/modal/ModalProvider.tsx:11]: Use better variable name here. state here is not context it's actual Modal component

File: src/shared/components/modal/ModalProvider.tsx (hunk 2 above)
Comment (sahil143) [reply src/shared/components/modal/ModalProvider.tsx:7]:
```
type ModalContextType = (hoc: (onClose: () => void) => React.ReactElement) => void;
```

File: src/shared/components/modal/createModalLauncher.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    <Modal
+      ref={null}
+      data-testId={`${props.title.toLowerCase().replace(' ', '-')}-modal`}
+      isOpen
+      onClose={handleClose}
+      aria-label={`${props.title} modal`}
+      {...props}
+    >
+      <Component {...props} onClose={handleClose} />
+    </Modal>
```
Comment (sahil143) [inline src/shared/components/modal/createModalLauncher.tsx:25]:
```suggestion
    <Modal
      data-testId={`${props.title.toLowerCase().replace(' ', '-')}-modal`}
      aria-label={`${props.title} modal`}
       {...props}
      isOpen
      onClose={handleClose}
      appendTo={() => document.querySelector('#hacDev-modal-container')}
    >
      <Component {...props} onClose={handleClose} />
    </Modal>
```

File: src/shared/components/modal/createModalLauncher.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+) => (props: P) => (onClose: () => void) => React.ReactElement;
+
+export const createModalLauncher: CreateModalLauncher = (Component) => (props) => (onClose) => {
+  const handleClose = () => {
+    props.onClose && props.onClose();
+    onClose();
+  };
+  return (
+    <Modal
+      ref={null}
```
Comment (sahil143) [inline src/shared/components/modal/createModalLauncher.tsx:17]: Why is ref set to null here?

File: src/shared/components/modal/modal-utils.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    setProps(null);
+  }, [setProps]);
+
+  const showModal = React.useCallback(
+    (launcherCallback) => {
+      const element = launcherCallback(handleClose);
+      setProps(element);
+    },
+    [handleClose, setProps],
+  );
```
Comment (sahil143) [inline src/shared/components/modal/modal-utils.ts:16]: Move these methods to the ModalProvider

File: src/shared/components/modal/modal-utils.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  return showModal;
```
Comment (sahil143) [inline src/shared/components/modal/modal-utils.ts:17]:
```suggestion
  const showModal = React.useContext(ModalContext);
  return showModal;
```

File: src/shared/components/modal/createModalLauncher.tsx [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react';
+import { Modal, ModalProps } from '@patternfly/react-core';
+
+export type ModalComponentProps = Omit<ModalProps, 'children'>;
```
Comment (sahil143) [inline src/shared/components/modal/createModalLauncher.tsx:4]:
```suggestion
export type ModalComponentProps = Omit<ModalProps, 'children' | 'isOpen' | 'appendTo'>;
```

File: src/components/LogViewer/ComponentLogViewerModal.scss [hunk 10]
```diff
@@ -1,3 +0,0 @@
-.hacDev-log-viewer-modal {
-  height: 70%;
-}
```
Comment (sahil143) [inline src/components/LogViewer/ComponentLogViewerModal.scss:3]: you need this CSS for Build log viewer

File: src/components/ApplicationDetailsView/ComponentListItem.tsx [hunk 11]
```diff
@@ -81,14 +78,18 @@ export const ComponentListItem: React.FC<ComponentListViewPageProps> = ({
...
               },
               {
-                cta: () => showLogsForComponent(component),
+                cta: () =>
+                  showModal(
+                    launchComponentLogViewerModal({
+                      component,
+                      title: 'View build logs',
+                    }),
+                  ),
```
Comment (sahil143) [inline src/components/ApplicationDetailsView/ComponentListItem.tsx:87]:
```suggestion
                  showModal(
                    launchComponentLogViewerModal({
                      component,
                      title: 'View build logs',
                      className: "hacDev-log-viewer-modal",
                      variant: ModalVariant.large
                    }),
                  ),
```

**Review by sahil143 — CHANGES_REQUESTED** [review]
@debsmita1 Do add some unit tests for ModalProvider and its utils.

File: src/shared/components/modal/createModalLauncher.tsx (hunk 6 above)
Comment (debsmita1) [reply src/shared/components/modal/createModalLauncher.tsx]: This was to fix a type error, removed it and handled the error

File: src/components/ApplicationDetailsView/ComponentListItem.tsx (hunk 11 above)
Comment (debsmita1) [reply src/components/ApplicationDetailsView/ComponentListItem.tsx]: I figured another way to pass these, I moved the Title inside the `BuildLogViewer` component that is rendered inside the Modal
and the remaining props that are needed for the modal I am now passing those as `args` along with the Component while calling the `createModalLauncher`. Let me know what you think about this

File: src/shared/components/modal/ModalProvider.tsx (hunk 4 above)
Comment (debsmita1) [reply src/shared/components/modal/ModalProvider.tsx]: changed it to `modal` and `setModal`

File: src/shared/components/modal/ModalProvider.tsx (hunk 1 above)
Comment (debsmita1) [reply src/shared/components/modal/ModalProvider.tsx]: missed removing this earlier, removed it now

**Comment by debsmita1** [comment]
working on tests

File: src/shared/components/modal/ModalProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+type ModalProviderProps = {
+  children: React.ReactNode;
+};
+
+export const ModalProvider: React.FC<ModalProviderProps> = ({ children }) => {
+  const [modal, setModal] = React.useState<React.ReactElement>();
+  const handleClose = React.useCallback(() => {
+    setModal(null);
+  }, [setModal]);
```
Comment (sahil143) [inline src/shared/components/modal/ModalProvider.tsx:16]: `setModal` don't need to be an dependency

File: src/shared/components/modal/ModalProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+  const showModal = React.useCallback(
+    (launcherCallback) => {
+      const element = launcherCallback(handleClose);
+      setModal(element);
+    },
+    [handleClose, setModal],
```
Comment (sahil143) [inline src/shared/components/modal/ModalProvider.tsx:23]: same as above

File: src/shared/components/modal/createModalLauncher.tsx [hunk 14]
```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
+import { Modal, ModalProps } from '@patternfly/react-core';
+
+export type ModalComponentProps = Omit<ModalProps, 'children' | 'isOpen' | 'appendTo'> & {
+  ref?: React.LegacyRef<Modal>;
+};
+
+export type CreateModalLauncher = <P extends ModalComponentProps>(
+  C: React.ComponentType<P>,
+  args?: {},
```
Comment (sahil143) [inline src/shared/components/modal/createModalLauncher.tsx:10]: Add a proper type for args, and these are going to be basically ModalComponentProps

File: src/shared/components/modal/createModalLauncher.tsx [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+) => (props: P) => (onClose: () => void) => React.ReactElement;
```
Comment (sahil143) [inline src/shared/components/modal/createModalLauncher.tsx:11]: Add a proper type for `(onClose: () => void) => React.ReactElement;` and use the same in ModalContext

File: src/shared/components/modal/modal-utils.ts [hunk 16]
```diff
@@ -0,0 +1,8 @@
+import * as React from 'react';
+import { ModalContext } from './ModalProvider';
+
+export const useModalLauncher = () => {
+  const showModal = React.useContext(ModalContext);
+
+  return showModal;
+};
```
Comment (sahil143) [inline src/shared/components/modal/modal-utils.ts:8]: you can move this to ModalProvider.tsx itself
```suggestion
export const useModalLauncher = () => React.useContext(ModalContext);
```

File: src/shared/components/modal/ModalProvider.tsx [hunk 17]
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import { ModalPortal } from '../../../ModalPortal';
+
+type ModalContextType = (hoc: (onClose: () => void) => React.ReactElement<any>) => void;
```
Comment (sahil143) [inline src/shared/components/modal/ModalProvider.tsx:4]: define a type for `(onClose: () => void) => React.ReactElement` and reuse it.

File: src/shared/components/modal/ModalProvider.tsx [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+
+export const ModalContext = React.createContext<ModalContextType>(() => {});
```
Comment (sahil143) [inline src/shared/components/modal/ModalProvider.tsx:6]: nit: No need to export this if you move `useModalLauncher` to this file.

**Review by sahil143 — COMMENTED** [review]
src/shared/components/modal/Modal.tsx
src/shared/components/modal/Modal.scss
src/shared/components/modal/index.ts

These three are empty files ^^ delete them

**Comment by debsmita1** [comment]
> @debsmita1 Do add some unit tests for ModalProvider and its utils.

Added tests. PTAL

File: src/shared/components/modal/__tests__/ModalProvider.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+      'data-testId': 'my-modal',
+    });
+  });
+  it('should not show modal', () => {
+    const TestComponent = () => {
+      const showModal = useModalLauncher();
+      return <Button onClick={() => showModal(launchMyModal({}))}>Show Modal</Button>;
+    };
+
+    const wrapper = mount(<ModalProvider>{TestComponent}</ModalProvider>);
```
Comment (sahil143) [inline src/shared/components/modal/__tests__/ModalProvider.spec.tsx:20]: it would be better if you could use react-testing-library for the test as we are using that for our tests in hac-dev.

File: src/components/LogViewer/BuildLogViewer.tsx [hunk 20]
```diff
@@ -1,15 +1,16 @@
...
 import { WatchK8sResource } from '../../dynamic-plugin-sdk';
+import { ModalComponentProps } from '../../shared/components/modal/createModalLauncher';
 import PipelineRunLogs from '../../shared/components/pipeline-run-logs/PipelineRunLogs';
 import { PipelineRunKind } from '../../shared/components/pipeline-run-logs/types';
 import { PipelineRunGroupVersionKind } from '../../shared/components/pipeline-run-logs/utils';
 import { EmptyBox, LoadingBox } from '../../shared/components/status-box/StatusBox';
 import { ComponentKind } from '../../types';
 
-type BuildLogViewerProps = {
+export type BuildLogViewerProps = ModalComponentProps & {
```
Comment (sahil143) [inline src/components/LogViewer/BuildLogViewer.tsx:13]: Why do you need to export this type?

File: src/components/NamespacedPage/NamespacedPage.tsx [hunk 21]
```diff
@@ -40,10 +41,12 @@ const NamespacedPage: React.FunctionComponent<NamespacedPageProps> = ({ children
...
-    <NamespaceContext.Provider value={{ namespace: activeNamepace }}>
-      <AppBanner />
-      <ErrorBoundary>{children}</ErrorBoundary>
-    </NamespaceContext.Provider>
+    <ModalProvider>
+      <NamespaceContext.Provider value={{ namespace: activeNamepace }}>
+        <AppBanner />
+        <ErrorBoundary>{children}</ErrorBoundary>
+      </NamespaceContext.Provider>
+    </ModalProvider>
```
Comment (rottencandy) [inline src/components/NamespacedPage/NamespacedPage.tsx:49]: I think modal provider should be inside the namespace provider, so components in modal could have access to the namespace context.

**Comment by rottencandy** [comment]
@rohitkrai03 so far the `src/shared` directory only contains components brought in from ODC, is this a convention we are following?

File: src/components/ApplicationDetailsView/ComponentListItem.tsx [hunk 22]
```diff
@@ -81,14 +78,17 @@ export const ComponentListItem: React.FC<ComponentListViewPageProps> = ({
                 label: 'Component settings',
               },
               {
-                cta: () => showLogsForComponent(component),
+                cta: () =>
```
Comment (rohitkrai03) [inline src/components/ApplicationDetailsView/ComponentListItem.tsx:81]: Probably better to pull out actions out of render now.

File: src/shared/components/modal/ModalProvider.tsx [hunk 23]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import { ModalPortal } from '../../../ModalPortal';
+import { ModalLauncher } from './createModalLauncher';
+
+type ModalContextType = (hoc: ModalLauncher) => void;
```
Comment (rohitkrai03) [inline src/shared/components/modal/ModalProvider.tsx:5]:
```suggestion
type ModalContextType = (launcher: ModalLauncher) => void;
```

File: src/shared/components/modal/ModalProvider.tsx [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+
+type ModalProviderProps = {
+  children: React.ReactNode;
+};
+
+export const useModalLauncher = () => React.useContext(ModalContext);
+
+export const ModalProvider: React.FC<ModalProviderProps> = ({ children }) => {
+  const [modal, setModal] = React.useState<React.ReactElement>();
+  const handleClose = React.useCallback(() => {
```
Comment (rohitkrai03) [inline src/shared/components/modal/ModalProvider.tsx:17]: Nit
```suggestion

  const handleClose = React.useCallback(() => {
```

File: src/shared/components/modal/__tests__/ModalProvider.spec.tsx [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+
+describe('Modal Provider', () => {
+  it('should launch modal on button click', async () => {
+    const launchMyModal = createModalLauncher(() => <div>My Modal</div>, {
+      'data-testid': 'my-modal',
+    });
+    const TestComponent = () => {
+      const showModal = useModalLauncher();
+      return (
+        <Button data-testid="show-modal-button" onClick={() => showModal(launchMyModal({}))}>
```
Comment (rohitkrai03) [inline src/shared/components/modal/__tests__/ModalProvider.spec.tsx:15]: Test fails if you pass null to `launchMyModal`. The launcher should not assume its going to get an object everytime.

File: src/shared/components/modal/createModalLauncher.tsx [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+export type ModalLauncher = (onClose: () => void) => React.ReactElement;
+
+export type CreateModalLauncher = <P extends ModalComponentProps>(
+  C: React.ComponentType<P>,
+  args?: ModalComponentProps & {
+    'data-testid': string;
+  },
+) => (props: P) => ModalLauncher;
```
Comment (rohitkrai03) [inline src/shared/components/modal/createModalLauncher.tsx:15]: Props shouldn't be a required field here so that we could use the modal in a use case where you don't need to pass down any component props.

File: src/shared/components/modal/createModalLauncher.tsx [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+import { Modal, ModalProps } from '@patternfly/react-core';
+
+export type ModalComponentProps = Omit<ModalProps, 'children' | 'isOpen' | 'appendTo'> & {
+  ref?: React.LegacyRef<Modal>;
+};
+
+export type ModalLauncher = (onClose: () => void) => React.ReactElement;
+
+export type CreateModalLauncher = <P extends ModalComponentProps>(
+  C: React.ComponentType<P>,
```
Comment (rohitkrai03) [inline src/shared/components/modal/createModalLauncher.tsx:11]:
```suggestion
  Component: React.ComponentType<P>,
```

File: src/shared/components/modal/createModalLauncher.tsx [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+  args?: ModalComponentProps & {
```
Comment (rohitkrai03) [inline src/shared/components/modal/createModalLauncher.tsx:12]: Probably better name to make it clear 
```suggestion
  modalConfig?: ModalComponentProps & {
```

File: src/shared/components/modal/createModalLauncher.tsx (hunk 26 above)
Comment (rohitkrai03) [inline src/shared/components/modal/createModalLauncher.tsx:15]:
```suggestion
) => (componentProps: P) => ModalLauncher;
```

File: src/shared/components/modal/createModalLauncher.tsx [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+  (Component, args) => (props) => (onClose) => {
+    const handleClose = () => {
+      props.onClose && props.onClose();
+      onClose();
+    };
+    return (
+      <Modal
+        aria-label="modal"
+        {...args}
+        {...props}
```
Comment (rohitkrai03) [inline src/shared/components/modal/createModalLauncher.tsx:27]: Why do we need to pass down component props to PF modal? Shouldn't `args` cover all the modal props?

File: src/shared/components/modal/createModalLauncher.tsx [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+  },
+) => (props: P) => ModalLauncher;
+
+export const createModalLauncher: CreateModalLauncher =
+  (Component, args) => (props) => (onClose) => {
+    const handleClose = () => {
+      props.onClose && props.onClose();
+      onClose();
+    };
+    return (
```
Comment (rohitkrai03) [inline src/shared/components/modal/createModalLauncher.tsx:23]:
```suggestion

    return (
```

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: src/shared/components/modal/ModalProvider.tsx [hunk 31]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (debsmita1) [reply src/shared/components/modal/ModalProvider.tsx:1]: @rohitkrai03 @rottencandy  where do you suggest moving these files to ? `src/components/` or simply under `src` ?

File: src/components/ApplicationDetailsView/launchComponentLogViewerModal.tsx [hunk 32]
```diff
@@ -0,0 +1,11 @@
+import { ModalVariant } from '@patternfly/react-core';
+import { createModalLauncher } from '../../shared/components/modal/createModalLauncher';
+import { BuildLogViewer } from '../LogViewer/BuildLogViewer';
+import './ComponentLogViewerModal.scss';
+
+export const launchComponentLogViewerModal = createModalLauncher(BuildLogViewer, {
```
[the remark this answers is not in the document]
Comment (debsmita1) [reply src/components/ApplicationDetailsView/launchComponentLogViewerModal.tsx]: Moved this inside BuildLogViewer component

File: src/components/LogViewer/BuildLogViewer.tsx [hunk 33]
```diff
@@ -76,3 +78,10 @@ export const BuildLogViewer: React.FC<BuildLogViewerProps> = ({ component }) =>
     </Stack>
   );
 };
+
+export const launchComponentLogViewerModal = createModalLauncher(BuildLogViewer, {
```
Comment (rohitkrai03) [inline src/components/LogViewer/BuildLogViewer.tsx:82]:
```suggestion
export const buildLogViewerLauncher = createModalLauncher(BuildLogViewer, {
```

File: src/components/modal/ModalProvider.tsx [hunk 34]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import { ModalPortal } from '../../ModalPortal';
```
Comment (rohitkrai03) [inline src/components/modal/ModalProvider.tsx:2]: Move `ModalPortal` to the same folder.

File: src/components/modal/createModalLauncher.tsx [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+
+export type CreateModalLauncher = <P extends ModalComponentProps>(
+  Component: React.ComponentType<P>,
+  modalConfig?: ModalComponentProps & {
+    'data-testid': string;
+  },
+) => (componentProps?: P) => ModalLauncher;
+
+export const createModalLauncher: CreateModalLauncher =
+  (Component, modalProps) => (props) => (onClose) => {
```
Comment (rohitkrai03) [inline src/components/modal/createModalLauncher.tsx:18]: Either keep it `modalConfig` or `modalProps` in both places. 
```suggestion
  (Component, modalProps) => (componentProps) => (onClose) => {
```

File: src/components/modal/createModalLauncher.tsx [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+export type ModalLauncher = (onClose: () => void) => React.ReactElement;
+
+export type CreateModalLauncher = <P extends ModalComponentProps>(
+  Component: React.ComponentType<P>,
+  modalConfig?: ModalComponentProps & {
+    'data-testid': string;
+  },
+) => (componentProps?: P) => ModalLauncher;
```
Comment (rohitkrai03) [inline src/components/modal/createModalLauncher.tsx:15]: Isn't `componentProps` just for passing data to actual component being rendered inside modal like `Component` data?

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

**Comment by rohitkrai03** [comment]
> src/shared/components/modal/Modal.tsx src/shared/components/modal/Modal.scss src/shared/components/modal/index.ts
> 
> These three are empty files ^^ delete them

Don't think this has been resolved.

File: src/shared/components/modal/ModalProvider.tsx (hunk 31 above)
[the remark this answers is not in the document]
Comment (rohitkrai03) [reply src/shared/components/modal/ModalProvider.tsx]: @debsmita1 We are now naming folders of components same as components. So maybe `src/components/Modal` or `src/components/ModalProvider` or `src/components/ModalLauncher`?

File: src/components/modal/createModalLauncher.tsx (hunk 36 above)
Comment (debsmita1) [reply src/components/modal/createModalLauncher.tsx:15]: @rohitkrai03  yes, it is
do you want me to name this to something else ? may be `componentData` ?

**Review by rohitkrai03 — APPROVED** [review]
/approve
/lgtm
